File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/data/ExclusiveSetGroup.java
package fr.hardel.asset_editor.client.javafx.lib.data;

import net.minecraft.resources.Identifier;

import java.util.List;

public record ExclusiveSetGroup(Identifier tagId) {

    public static final List<ExclusiveSetGroup> ALL = List.of(
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/armor")),
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/bow")),
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/crossbow")),
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/damage")),
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/riptide")),
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/mining")),
            new ExclusiveSetGroup(Identifier.fromNamespaceAndPath("minecraft", "exclusive_set/boots"))
    );

    public String id() {
        return tagId.getPath();
    }

    public Identifier image() {
        return tagId.withPath("textures/studio/enchantment/" + tagId.getPath() + ".png");
    }

    public String value() {
        return "#" + tagId;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/data/ExclusiveSetGroup.kt
package fr.hardel.asset_editor.client.compose.lib.data

import fr.hardel.asset_editor.client.AssetEditorClient
import net.minecraft.resources.Identifier

data class ExclusiveSetGroup(val tagId: Identifier) {

    fun id(): String = tagId.path

    fun image(): Identifier =
        tagId.withPath("textures/studio/enchantment/${tagId.path}.png")

    fun value(): String = "#$tagId"

    companion object {
        private val EXCLUSIVE_GROUP = Identifier.fromNamespaceAndPath("asset_editor", "exclusive")

        val ALL: List<ExclusiveSetGroup>
            get() = AssetEditorClient.studioConfigMemory().snapshot()
                .enchantmentEntriesFor(EXCLUSIVE_GROUP)
                .map { entry -> ExclusiveSetGroup(entry.id()) }
    }
}
